### PR TITLE
Warnings for `two_qubit_unitary`

### PR DESCRIPTION
**Pull request by JerryChen97**
**Context:**
`qml.ops.op_math.decompositions.two_qubit_unitary.py` is not differentiable.
**Description of the Change:**
Will raise warning if users try call this method in differentiable way
**Benefits:**
Raise caution regarding the non-differentiability of affected program
Temporarily help prevent usage of this method in differentiable way until full implementation in the future
**Possible Drawbacks:**
Not fully fix the affected issue.
**Related GitHub Issues:**
#6045 

**Related Shortcut Stories:**
[sc-69959]

### CODE REVIEW

**Comment by JerryChen97** [comment]
The replacement of the `eigvals` call with shorter, conciser condition tester during the branching within `_compute_num_cnots`  won't help too much over the original issue. The only effect is there will be no more `NotImplementedError`, but the overall non-differentiability issue persists.

File: pennylane/ops/op_math/decompositions/two_qubit_unitary.py [hunk 1]
```diff
@@ -43,6 +45,46 @@
...
+            isinstance(U, tf.Tensor) and not U.dtype.is_floating
+        )
+    elif interface == "jax":
+        # For JAX, consider arrays in a quantum context as potentially trainable
+        is_trainable = True
+    elif interface == "autograd":
+        # For autograd, arrays in QNodes are potentially trainable
+        is_trainable = True
+
+    if is_trainable:
```
Comment (albi3ro) [inline pennylane/ops/op_math/decompositions/two_qubit_unitary.py:79]: So we do actually have `qml.math.requires_grad` that does the above logic.
```suggestion
    if qml.math.requires_grad(U):
```

**Review by albi3ro — APPROVED** [review]
:+1

File: pennylane/ops/op_math/decompositions/two_qubit_unitary.py [hunk 2]
```diff
@@ -43,6 +45,22 @@
 ###################################################################################
 
 
+def _check_differentiability_warning(U):
+    """Check conditions that may lead to non-differentiability and raise appropriate warnings.
+
+    Args:
+        U (tensor): Input unitary matrix to check
```
Comment (PietropaoloFrisoni) [inline pennylane/ops/op_math/decompositions/two_qubit_unitary.py:52]:
```suggestion
        U (tensor_like): Input unitary matrix to check.
```
I think this is the 'usual' PL way to indicate tensor-like args (it is just the notation I have seen more often, but I am not sure)

**Review by PietropaoloFrisoni — APPROVED** [review]
Looks great to me! Just a totally optional suggestion

**Comment by JerryChen97** [comment]
Thanks a lot for reviews! I'll mod the changelog and merge

**Pull request merged**